### DISCUSSION TITLE
Hover working for modules but not classes/functions for internal libraries

### DISCUSSION THREAD

**Discussion by ngriffiths13** [zed-industries/zed#6623] [Feature Requests]
I am not actually sure if this is an issue with Pyright or with Zed, but I thought I would see if anyone could point me in the right direction here.

When hovering over imported python modules (from internally hosted private pypi) I can see all of the docstrings and things I would expect to see. However, when I hover over the classes or functions I am importing from those modules, it shows up as `Unknown`. This doesn't happen when I install packages from the public PyPI repo.

Anyone have any idea why this might be happening or where I can start poking around to find a fix?

**Comment by JosephTLyons** [comment]
Hey @ngriffiths13, do other Pyright features for you work, such as autocompletion?

**Reply by ngriffiths13** [reply]
@JosephTLyons the other pyright features work for other packages. The autocompletion even works on import statements until I try to import a class or function. So in the example:
``` from my_package.my_module import my_function```
All pyright features work for `my_package` and `my_module` but not for `my_function`

**Reply by ngriffiths13** [reply]
I should also mention that in the past, these same functions and classes worked in vscode and pycharm. So I don't think it is the way we have our packages structured?

**Reply by JosephTLyons** [reply]
Hmm, interesting.  I'm not able to reproduce this in my local projects.  If you have the time, would you be able to zip up a project that exhibits the issue, preferably the smallest amount of code necessary to reproduce the issue? This issue seems strange.

**Reply by ngriffiths13** [reply]
I am not quite sure how to provide you a reproducible example. The issue only arises with private packages my company has built and maintains, which I can't share. Any public libraries work just as expected. Maybe I could build and push a small example library, install it, and zip up that venv for you to see. Give me a day and Ill see what I can do.

**Reply by ngriffiths13** [reply]
@JosephTLyons I think this should do it. This setup is causing me the same issue.
[hello-world.zip](https://github.com/zed-industries/community/files/12524654/hello-world.zip)

**Reply by ngriffiths13** [reply]
@JosephTLyons were you able to reproduce what I was seeing with the zip file?

**Reply by JosephTLyons** [reply]
Sorry @ngriffiths13, got lost in a few other things.  Will check this out shortly.  Thank you for putting this together.

**Reply by ngriffiths13** [reply]
No worries! I know you are busy with a lot of things. Not stopping me from using Zed at all. So whenever you have time is fine. Thank you again

**Reply by JosephTLyons** [reply]
Ahh, I see what you're saying now.  Something interesting I saw was that `say_hello` wasn't being picked up initially, but if I navigated to `my_module` by command+clicking on it, then went back to the `__init__.py` file, `say_hello` was then recognized by Pyright. Curious what VS Code / Pylance is doing to circumvent this issue.

**Reply by ngriffiths13** [reply]
Ah that is interesting. Well now I at least have a workaround 😅. So it seems like a pyright problem but one you might want to fix on the Zed side?

**Reply by ngriffiths13** [reply]
Unfortunately that workaround actually requires me going into the source files of every single class and function I want to import and not just to the top level module. I don't know that much about how pyright works. But if you have any other ideas for less time consuming workarounds right now I'd be happy to hear them!

**Reply by ngriffiths13** [reply]
I just found something else interesting. When navigating to the source file for classes, Pyright only recognizes classes and functions while I have that file open. Once I close that file again, it goes back to an Unknown object.

**Reply by ngriffiths13** [reply]
@JosephTLyons just wanted to check in to see if there is anything can be done here about this?

**Reply by JosephTLyons** [reply]
Sorry for the late reply @ngriffiths13, Ill bring it up the with the team, but it might fall into the category of "after we go open source," which is imminent.  We want to be careful what all blocks the work on open source because opening the code base allows anyone to come in and fix this stuff! Hang tight.

**Reply by ngriffiths13** [reply]
Thanks for getting back on this! I totally understand. Looking forward to your open source launch!

**Comment by jeff-phil** [comment]
Related?:  https://github.com/microsoft/pyright/issues/2135

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
